sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeArrayData.java

```diff
@@ -452,6 +452,16 @@ public UnsafeArrayData copy() {
 
   public static UnsafeArrayData fromPrimitiveArray(
        Object arr, int offset, int length, int elementSize) {
+    UnsafeArrayData result = forPrimitiveArray(length, elementSize);
+    final long headerInBytes = calculateHeaderPortionInBytes(length);
+    final long valueRegionInBytes = (long)elementSize * length;
+    final Object data = result.getBaseObject();
+    Platform.copyMemory(arr, offset, data,
+      Platform.LONG_ARRAY_OFFSET + headerInBytes, valueRegionInBytes);
+    return result;
+  }
+
+  public static UnsafeArrayData forPrimitiveArray(int length, int elementSize) {
```

Comment:
how about `createFreshArray`? It's not for primitive only, e.g. we use it to store a struct of fixed size fields.